File: advent_of_code_2025/day05/Main.hs
{-# LANGUAGE OverloadedStrings #-}

module Main where

import Data.Attoparsec.Text
import Data.Text (Text)
import Data.Text.IO qualified as T

parse_ :: Text -> ([(Int, Int)], [Int])
parse_ = either (error "Bad parse") id . parseOnly twoSections
  where
    twoSections = (,) <$> (range `sepBy1'` char '\n') <*> ("\n\n" *> (ingredient `sepBy1'` char '\n'))
    range = (,) <$> decimal <*> (char '-' *> decimal)
    ingredient = decimal

part1 :: [(Int, Int)] -> [Int] -> Int
part1 ranges = length . filter (isFresh ranges)
  where
    isFresh rs i = any (\(lo, hi) -> lo <= i && i <= hi) rs

main :: IO ()
main = do
    (ranges, ingredients) <- parse_ <$> T.readFile "input.txt"
    print $ part1 ranges ingredients

File: advent_of_code_2025/day05/src/main.rs
use rayon::prelude::*;
use std::collections::BTreeSet;

fn parse(s: &str) -> (Vec<(u64, u64)>, Vec<u64>) {
    let (first, second) = s.split_once("\n\n").expect("spacing");
    let ranges = first
        .trim()
        .lines()
        .map(|line| {
            let (a, b) = line.trim().split_once('-').expect("hyphen");
            (a.parse().expect("a"), b.parse().expect("b"))
        })
        .collect();
    let ingredients = second
        .trim()
        .lines()
        .map(|line| line.trim().parse().expect("ingredient"))
        .collect();
    (ranges, ingredients)
}

fn part1(ranges: &[(u64, u64)], ingredients: &[u64]) -> usize {
    ingredients
        .iter()
        .filter(|&&i| ranges.iter().any(|&(a, b)| a <= i && i <= b))
        .count()
}

fn part2(ranges: &[(u64, u64)]) -> usize {
    ranges
        .into_par_iter()
        .fold_chunks(5, BTreeSet::new, |mut acc, &(a, b)| {
            for n in a..=b {
                acc.insert(n);
            }
            acc
        })
        .fold(BTreeSet::new, |mut acc, xs| {
            for n in xs {
                acc.insert(n);
            }
            acc
        })
        .count()
}

fn main() {
    let (ranges, ingredients) = parse(&std::fs::read_to_string("input.txt").expect("file"));
    println!("{}", part1(&ranges, &ingredients));
    //println!("{}", part2(&ranges));
}
